File: solutions/python/bob/1/bob.py
def response(hey_bob):
    is_question = hey_bob.rstrip().endswith('?')
    is_yell = hey_bob == hey_bob.upper() != hey_bob.lower()
    is_silence = hey_bob.isspace() or not hey_bob
    if is_question and is_yell:
        result = 'Calm down, I know what I\'m doing!'
    elif is_question:
        result = 'Sure.'
    elif is_yell:
        result = 'Whoa, chill out!'
    elif is_silence:
        result = 'Fine. Be that way!'
    else:
        result = 'Whatever.'
    return result
    

File: solutions/erlang/bob/1/src/bob.erl
-module(bob).

-export([response/1]).

response(String) ->
    response_empty(is_empty(String), String).

response_empty(true, _) -> "Fine. Be that way!";
response_empty(false, UntrimmedString) ->
    String = string:trim(UntrimmedString), 
    response_yell(is_yell(String), String).

response_yell(true, String) ->
    response_yell_if_question(is_question(String));
response_yell(false, String) ->
    response_question(is_question(String)).

response_yell_if_question(false) -> "Whoa, chill out!";
response_yell_if_question(true) -> "Calm down, I know what I'm doing!".

response_question(true) -> "Sure.";
response_question(false) -> "Whatever.".

is_yell(String) -> 
    String =:= string:to_upper(String) andalso
    String =/= string:to_lower(String).

is_question(_String) ->
    $? == lists:last(_String).

is_empty([]) -> true;
is_empty(String) ->
    string:trim(String) == [].